xtask/codegen/gritql.ungram

```diff
@@ -0,0 +1,554 @@
+// GritQL Un-Grammar.
+//
+// This grammar specifies the structure of Rust's concrete syntax tree.
+// It does not specify parsing rules (ambiguities, precedence, etc are out of scope).
+// Tokens are processed -- contextual keywords are recognised, compound operators glued.
+//
+// Legend:
+//
+//   //          				-- comment
+//   Name =      				-- non-terminal definition
+//   'ident'     				-- token (terminal)
+//   A B         				-- sequence
+//   A | B       				-- alternation
+//   A*          				-- zero or more repetition
+//   (A (',' A)* ','?)	        -- repetition of node A separated by ',' and allowing a trailing comma
+//   (A (',' A)*)	            -- repetition of node A separated by ',' without a trailing comma
+//   A?          				-- zero or one repetition
+//   (A)         				-- same as A
+//   label:A     				-- suggested name for field of AST node
+
+// NOTES
+//
+// - SyntaxNode, SyntaxToken and SyntaxElement will be stripped from the codegen
+// - Bogus nodes are special nodes used to keep track of broken code; they are
+//   not part of the grammar but they will appear inside the green tree
+//
+
+
+///////////////
+// Bogus NODES
+///////////////
+// SyntaxElement is a generic data structure that is meant to track nodes and tokens
+// in cases where we care about both types
+//
+// As Bogus* node will need to yield both tokens and nodes without discrimination,
+// and their children will need to yield nodes and tokens as well.
+// For this reason, SyntaxElement = SyntaxElement
+SyntaxElement = SyntaxElement
+
+GritBogus = SyntaxElement*
+GritBogusContainer = SyntaxElement*
+GritBogusDefinition = SyntaxElement*
+GritBogusPattern = SyntaxElement*
+GritBogusLiteral = SyntaxElement*
+GritBogusNamedArg = SyntaxElement*
+GritBogusPredicate = SyntaxElement*
+
+GritRoot =
+    bom: 'UNICODE_BOM'?
+    version: GritVersion?
+    language: GritLanguageDeclaration?
+    definitions: GritDefinitionList?
+    pattern: AnyGritPattern?
+    definitions_continued: GritDefinitionList?
+    eof: 'EOF'
+
+GritSequential = 'sequential' '{' sequential: GritPatternList '}'
+
+GritFiles = 'multifile' '{' files: GritPatternList '}'
+
+AnyGritDefinition =
+    GritPatternDefinition
+    | GritPredicateDefinition
+    | GritFunctionDefinition
+    | GritBogusDefinition
+
+GritDefinitionList = AnyGritDefinition ('newline' AnyGritDefinition)* 'newline'?
+
+GritVersion = 'engine' 'biome' '(' GritDoubleLiteral ')'
+
+GritLanguageDeclaration =
+    'language'
+    name: GritLanguageName
+    flavor: GritLanguageFlavor?
+    GritBogus?
```

Comment:
A bogus should not be in this position. Usually bogus nodes are part of a `Any**` node.

Reply:
It helped me to implement patterns such as this in the parser:

```rs
let _ = parse_pattern(p).or_recover_with_token_set(
    p,
    &ParseRecoveryTokenSet::new(GRIT_BOGUS_PATTERN, PATTERN_RECOVERY_SET),
    expected_pattern,
);
```

Without this, it appeared as though the bogus nodes wouldn't automatically "bubble up" to the next `Any**` node. I suppose that might simply be because I had made some other mistake, what steps do I need to follow to make sure the `Any*` nodes can consume these?

Reply:
There aren't particular steps to follow, really. The recovering logic is up to us entirely, and we decide **what** to wrap in bogus node. We just make judgement calls based on the DX we want to offers to our tools (formatter, linter, etc.).

By looking at the snippet of the parser, I would expect a grammar like this:

```
GritNode =
	some_node: GritSomeNode
	pattern: AnyGritPattern

AnyGritPattern =
	GritBogusPattern
	GritSpecificPattern
```

This means when you're parsing the `pattern` node, you have to account any possible errors, and make sure the parser is able to recover itself and emit a `GRIT_BOGUS_PATTERN`.

Reply:
Thanks, maybe I was just trying to do the recovery in the wrong node then.

Reply:
If you have questions, @denbezrukov is an expert in the field :)